Update agent eval tutorial (#582)
langchain-ai/langsmith-docs: docs/evaluation/tutorials/agents.mdx

import { RegionalUrl } from "@site/src/components/RegionalUrls";

# Evaluate an agent

:::info Key concepts
[Agent evaluation](../concepts#agents) | [Evaluators](../concepts#evaluators) | [LLM-as-judge evaluators](../concepts#llm-as-judge)
:::

> In this tutorial, we'll build a customer support bot that helps users navigate a digital music store. We'll create three types of evaluations:

Reviewer: ```suggestion
In this tutorial, we'll build a customer support bot that helps users navigate a digital music store. Then, we'll go through the three most effective types of evaluations to run on chat bots:
```

- [Final response](../concepts#evaluating-an-agents-final-response): Evaluate the agent's final response.
- [Trajectory](../concepts#evaluating-an-agents-trajectory): Evaluate whether the agent took the expected path (e.g., of tool calls) to arrive at the final answer.
- [Single step](../concepts#evaluating-a-single-step-of-an-agent): Evaluate any agent step in isolation (e.g., whether it selects the appropriate first tool for a given ).

We'll build our agent using [LangGraph](https://github.com/langchain-ai/langgraph), but the techniques and LangSmith functionality shown here are framework-agnostic.

## Setup

### Configure the environment

Let's install the required dependencies:

```bash
pip install -U langgraph langchain langchain-community langchain-openai
```

> and set up our environment variables for OpenAI and <RegionalUrl text='LangSmith'/>:

Reviewer: ```suggestion
Let's set up environment variables for OpenAI and <RegionalUrl text='LangSmith'/>:
```

```python
#region
import getpass
import os

def _set_env(var: str) -> None:
    if not os.environ.get(var):
        os.environ[var] = getpass.getpass(f"Set {var}: ")

os.environ["LANGCHAIN_TRACING_V2"] = "true"
_set_env("LANGCHAIN_API_KEY")
_set_env("OPENAI_API_KEY")
#endregion
```

> 

Reviewer: While we're using OpenAI for this example, you can use any [LangChain supported chat model](https://python.langchain.com/docs/integrations/chat/#all-chat-models).

> 

Reviewer: ```suggestion
While we're using OpenAI for this example, you can use any [LangChain supported chat model](https://python.langchain.com/docs/integrations/chat/#all-chat-models).
```

Author: i dont know if we should even say that you need to use langchain. yes this code is langchain-based but you could apply all the same techniques without it, from langsmith perspective it doesn't matter if you use langchain

    ]
#endregion

# Model for performing extraction.
info_llm = init_chat_model("gpt-4o-mini").with_structured_output(
    PurchaseInformation, method="json_schema", include_raw=True
)

# Graph node for extracting user info and routing to lookup/refund/END.
#region
async def gather_info(state: State) -> Command[Literal["lookup", "refund", END]]:
    info = await info_llm.ainvoke(
        [
            {"role": "system", "content": gather_info_instructions},
            *state["messages"],
        ]
    )
    parsed = info["parsed"]
    if any(parsed[k] for k in ("invoice_id", "invoice_line_ids")):
        goto = "refund"
    elif all(
        parsed[k]
        for k in ("customer_first_name", "customer_last_name", "customer_phone")
    ):
        goto = "lookup"
    else:
        goto = END
    update = {"messages": [info["raw"]], **parsed}
    return Command(update=update, goto=goto)
#endregion

# Graph node for executing the refund.
# Note that here we inspect the runtime config for an "env" variable.
# If "env" is set to "test", then we don't actually delete any rows from our database.

> # This will become important when we're runnign our evaluations.

Reviewer 2: typo

Author: ```suggestion
# This will become important when we're running our evaluations.
```

    ]

# Model for performing extraction.
info_llm = init_chat_model("gpt-4o-mini").with_structured_output(
    PurchaseInformation, method="json_schema", include_raw=True
)

# Graph node for extracting user info and routing to lookup/refund/END.
async def gather_info(state: State) -> Command[Literal["lookup", "refund", END]]:
    info = await info_llm.ainvoke(
        [
            {"role": "system", "content": gather_info_instructions},
            *state["messages"],
        ]
    )
    parsed = info["parsed"]
    if any(parsed[k] for k in ("invoice_id", "invoice_line_ids")):
        goto = "refund"
    elif all(
        parsed[k]
        for k in ("customer_first_name", "customer_last_name", "customer_phone")
    ):
        goto = "lookup"
    else:
        goto = END
    update = {"messages": [info["raw"]], **parsed}
    return Command(update=update, goto=goto)

# Graph node for executing the refund.
# Note that here we inspect the runtime config for an "env" variable.
# If "env" is set to "test", then we don't actually delete any rows from our database.

> # This will become important when we're runnign our evaluations.

Reviewer 2: typo

Author: ```suggestion
# This will become important when we're running our evaluations.
```
